src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt

```diff
@@ -42,25 +43,45 @@ internal class MessageReceiverExample : MessageReceiver {
     }
 }
 
+class SubscribeManager(subscriptionId: String) {
+    private val dockerProcessor = DockerProcessor()
+    private val subscriptionName = SubscriptionName.of(PROJECT_ID, subscriptionId)
+    private var subscriber: Subscriber? = null
+
+    fun subscribe(callback: (message: String, md5sum: String?) -> Unit) {
+        try {
+            subscriber = Subscriber.newBuilder(subscriptionName, MessageReceiverExample()).build()
+            subscriber?.startAsync()?.awaitRunning()
+
+            while (true) {
+                processMessage(callback)
+            }
+        } finally {
+            subscriber?.stopAsync()
+        }
+    }
+
+    fun processMessage(callback: (message: String, md5sum: String?) -> Unit) {
```

Comment:
I am not sure if readFully() can actually return null or not, but even if it can, you can check it in your DockerProcessor and return non-null. And remove nullability from `md5sum`

Reply:
Do I have to return non-null besides try-catch in docker processor?

Reply:
Ah, you have a try-catch block in the processor, sorry. In my opinion it is okay to rethrow the exception from catch. In the DockerProcessor you can't do anything meaningful with the exception besides logging, while caller has to process exceptions anyway, so it will probably be more happier with exception as error indicator rather than with null valuye

Reply:
Created exception after try-catch block. Done